### PR TITLE
86b1tvpke:Add and manage students activity log

### PR DESCRIPTION
**Pull request by LakshananS**
**Backend Updates - HTTP Methods for Activity Management**

**Implemented HTTP Methods**
Added functionality for handling CRUD operations on activities. The following methods are now available:
Create: - Added support for creating new activities.
Delete - Implemented the ability to delete activities.
Update -  Enabled updating of existing activities with validation to restrict modifications beyond 2 days from creation.
View - Added functionality to retrieve activity details.
Updated Middleware - Enhanced middleware to manage HTTP methods effectively, including restrictions on editing and deleting activities based on age.

**JWT Authentication** 
Integrated JSON Web Token (JWT) for secure user authentication, ensuring that only authorized users can perform certain actions.

**Prisma Singleton** 
Implemented a Prisma Client singleton to optimize database interactions and ensure consistent client usage across the application.

I installed following dependencies for perform the tasks:
* install Prisma client
* install jwctoken 
* npm i --save-dev @types/jsonwebtoken

### CODE REVIEW

File: server/src/middleware.ts [hunk 1]
```diff
@@ -1,19 +1,30 @@
-import { NextRequest, NextResponse } from "next/server";
+import { NextRequest, NextResponse } from 'next/server';
+import jwt from 'jsonwebtoken';
 
 export function middleware(request: NextRequest) {
-    console.log("Middleware called");
+    // Skip token validation for API routes
```
Comment (bsaranga) [inline server/src/middleware.ts]: Why aren't we validating tokens for api routes?

File: server/.env [hunk 2]
```diff
@@ -7,4 +7,5 @@ BASE_URL=http://localhost:3000
 # Prisma supports the native connection string format for PostgreSQL, MySQL, SQLite, SQL Server, MongoDB and CockroachDB.
 # See the documentation for all the connection string options: https://pris.ly/d/connection-strings
 
-DATABASE_URL="postgresql://postgres:postgres@localhost:5432/postgres?schema=public"
\ No newline at end of file
+DATABASE_URL="postgresql://postgres:postgres@localhost:5432/postgres?schema=public"
+JWT_SECRET= "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
```
Comment (bsaranga) [inline server/.env]: This looks like a jwt token itself. The secret can be a short random string.

File: server/src/app/api/blogs/[id]/route.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
+import { NextRequest, NextResponse } from "next/server";
+import { PrismaClient } from '@prisma/client';
+
+const prisma = new PrismaClient();
```
Comment (bsaranga) [inline server/src/app/api/blogs/[id]/route.ts]: Use the Prisma client singleton without re-instantiating

File: server/src/middleware.ts [hunk 4]
```diff
@@ -1,19 +1,30 @@
...
-    if (!cookie || cookie.value != 'authenticated') {
+    if (!cookie || !isValidToken(cookie.value)) {
         return NextResponse.redirect(`${process.env.BASE_URL}/login`);
     }
     return NextResponse.next();
 }
 
+const isValidToken = (token: string) => {
+    try {
+        const decoded = jwt.verify(token, process.env.JWT_SECRET || 'your-default-secret');
```
Comment (bsaranga) [inline server/src/middleware.ts]: Let's not assume that `JWT_SECRET` is nullish. It should be supplied for the application to work. Remove `|| 'your-default-secret'`

**Review by bsaranga — CHANGES_REQUESTED** [review]

File: server/.env (hunk 2 above)
Comment (LakshananS) [reply server/.env]: If I use the following command to generate the key, will it solve the problem?
node -e "console.log(require('crypto').randomBytes(64).toString('hex'))"

**Pull request closed without merging**